test/es-module/test-esm-loader-hooks.mjs

```diff
@@ -0,0 +1,79 @@
+// Flags: --expose-internals
+import { mustCall } from '../common/index.mjs';
+import esmLoaderModule from 'internal/modules/esm/loader';
+import assert from 'assert';
+
+const { ESMLoader } = esmLoaderModule;
+
+/**
+ * Verify custom hooks are called with appropriate arguments.
+ */
+{
+  const esmLoader = new ESMLoader();
+
+  const originalSpecifier = 'foo/bar';
+  const importAssertions = { type: 'json' };
+  const parentURL = 'file:///entrypoint.js';
+  const resolvedURL = 'file:///foo/bar.js';
+  const suggestedFormat = 'test';
+
+  function resolve(specifier, context, defaultResolve) {
+    assert.strictEqual(specifier, originalSpecifier);
+    // Ensure `context` has all and only the properties it's supposed to
+    assert.deepStrictEqual(Object.keys(context), [
+      'conditions',
+      'importAssertions',
+      'parentURL',
+    ]);
+    assert.ok(Array.isArray(context.conditions));
+    assert.strictEqual(context.importAssertions, importAssertions);
```

Comment:
I suggest using `JSON.stringify` here to compare both objects
```suggestion
    assert.strictEqual(JSON.stringigy(context.importAssertions), JSON.stringify(importAssertions));
```

Reply:
okay! FMI, why is that better?

Reply:
We don't want the object to be reference-equal since we are letting the hooks returning whatever `importAssertion` object, but we want both object to have the same content. Now that I think about it, maybe an even better way would be to use `deepStrictEqual` instead of `strictEqual`+`JSON.stringify` btw.

Reply:
Oh sweet, that's what I had locally. Was about to check it in til you posted the stringify. Incoming